[CLIv2] Initial commit of generating server side completions

jamesls opened the pull request:
This commit includes only the heuristics for generating
these completions.  It does not include the actual generated
modules.  That will be a large commit so I want to hold off on
sending that until this commit is reviewed.  Changes to this
script could change what we generate.

kyleknap commented on awscli/autocomplete/autogen.py:
```diff
@@ -0,0 +1,177 @@
+    """Simple implementation of making a word singular.
+
+    Where possible, you should use nltk.stem.WordNetLemmatizer.
+    It's not included here as a default to avoid a hard dependenct
```
dependenct -> dependency

kyleknap commented on awscli/autocomplete/autogen.py:
```diff
@@ -0,0 +1,177 @@
+            if op_name.lower().startswith(prefix):
+                op_with_prefix_removed = op_name[len(prefix):]
+                return op_with_prefix_removed
+                break
```
I don't think the `break` statement is needed here.

kyleknap commented on awscli/autocomplete/autogen.py:
```diff
@@ -0,0 +1,177 @@
+        list_member = output.members[resource_shape_name].member
+        op_with_prefix_removed = self._remove_verb_prefix(op_name)
+        if list_member.type_name == 'structure':
+            candidates = list_member.members
```
It might be good to break these if statements into separate internal methods for readability

kyleknap commented on awscli/autocomplete/autogen.py:
```diff
@@ -0,0 +1,177 @@
+class ServerCompletionHeuristic(object):
+    # Prefix that typically indicates listing resources,
+    # e.g ListResources, DescribeResources, etc.
+    _RESOURCE_VERB_PREFIX = ('list', 'describe')
```
We may need to add `get`, but not sure how that affects the results. I know some services such as API Gateway use `get` instead of `list` and `describe`

jamesls replied:
I went ahead and added it.  It's not great but it's better than nothing.  Here's an example of api gateway: https://gist.github.com/jamesls/043bb24bd2a6efd427109f72be2d7eca

jamesls replied:
Actually, I'll have to remove this for now.  Right now we make the assumption that the operations fall into two buckets, one for auto completing server side values, and the other for operations that *can* be auto-completed.  `Get` can be used for both cases so until there's more sophisticated logic here it can't be part of `_RESOURCE_VERB_PREFIX`.

kyleknap commented on awscli/autocomplete/autogen.py:
```diff
@@ -0,0 +1,177 @@
+    def _generate_operations(self, op_names, resources, service_model):
+        # member_name -> resource
+        reverse_mapping = {}
+        for name, identifier_map in resources.items():
```
It would be nice to break this up into smaller internal methods for readability as well

kyleknap commented on awscli/autocomplete/autogen.py:
```diff
@@ -0,0 +1,177 @@
+        output = op_model.output_shape
+        list_members = [member for member, shape in output.members.items()
+                        if shape.type_name == 'list']
+        if len(list_members) != 1:
```
How again do we model resources that have a required parameter in order to list (i.e. subresource)?

jamesls replied:
Via the `parameters` key.  Right now we always generate an empty map so we're not taking advantage of this feature.

jamesls commented:
@kyleknap Ok feedback incorporated.  I also added a new file `scripts/gen-server-completions` that's the main driver for generating these files.  I'll send a follow up PR for the generated files once we get this merged, it's a pretty large diff.

jamesls commented:
To simplify things for now, I added the inputParameters to the resources when they have required params, but I removed any operations that use a resource that has required params.  I expect to come back to this and try to auto-gen required params for operations, but for now this keeps things simple.

cc @kyleknap
